### PR TITLE
Add MSCOCO_2017 dataset for single pose

### PR DESCRIPTION
**Pull request by yd8534976**
- KeypointDetectionBase class
- MscocoSinglePersonKeypoints class
- Related minor fixes
- Related pre processor
- Related unit tests
- MSCOCO_2017 fixture dataset

## How has this been tested?

Only a unit test for dataset loader is not intuitive.
This branch #386 will be helpful to review.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature / Optimization (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by yd8534976** [comment]
@ruimashita 
I'm sorry for inconvenience, I have extracted dataset loader part from branch #386 .

File: lmnet/tests/check_dataset_storage.py [hunk 1]
```diff
@@ -742,6 +743,54 @@ def test_mscoco_object_detection_person():
         assert labels.shape[2] == 5
 
 
+def test_mscoco_2017_single_pose_estimation():
```
[the remark this answers is not in the document]
Comment (tkng) [reply lmnet/tests/check_dataset_storage.py]: That's a difficult question... In short, yes, we're maintaining `check_dataset_storage.py`. However, the question is that maintaining it is good idea? I'm not sure...

For me, just adding an unit test file to `lmnet/tests/lmnet_tests/datasets_tests/` looks enough. But I'd like to know the opinion of @ruimashita.

[the remark this answers is not in the document]
Comment (ruimashita) [reply lmnet/tests/check_dataset_storage.py]: `check_dataset_storage.py`, the main purpose of this script is to check the each datasets is stored correct on storage of servers. I believe that to confirm datasets files is productive for guarantee model's results, but current way seems not good cause it takes long time. Maybe We need to change this design to something like hash checking.
Also we need to implement unit test for each lmnet.datasets.xxx class, this should implement in `lmnet/tests/lmnet_tests/datasets_tests/`.
My opinion of test of dataset is 
* Need to implement unit test in `lmnet/tests/lmnet_tests/datasets_tests/`.
* Not need but recommend to implement test in `check_dataset_storage.py`. But this script will be drastically change.

File: lmnet/lmnet/pre_processor.py [hunk 2]
```diff
@@ -213,6 +240,79 @@ def _liner_quantize(x, bit, value_min, value_max):
     return result * value_range + value_min
 
 
+def joints_to_gaussian_heatmap(joints, image_size,
```
[the remark this answers is not in the document]
Comment (yd8534976) [reply lmnet/lmnet/pre_processor.py]: I have added unit tests and fixture dataset. please review. 🙇

File: lmnet/tests/lmnet_tests/test_pre_processor.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+                             [16, 16, 1],
+                             [17, 17, 0]])
+
+    pre_process = JointsToGaussianHeatmap(image_size, num_joints=num_joints,
+                                          stride=stride, sigma=2, max_value=10)
+
+    heatmap = pre_process(joints=input_joints)["heatmap"]
+
+    # It is hard to test semantic correctness of a gaussian heatmap manually.
+    # That part will be tested jointly with FormatJoints() in test_post_processor.py.
```
Comment (ruimashita) [inline lmnet/tests/lmnet_tests/test_pre_processor.py]: Will you implement `FormatJoints()` in another PR ?

Comment (yd8534976) [reply lmnet/tests/lmnet_tests/test_pre_processor.py]: @ruimashita Yes! I'll make another PR for post processing.
Test will be like:
https://github.com/blue-oil/blueoil/blob/f0f3b821a9ba91ee4fa511cfa70b70474d145899/lmnet/tests/lmnet_tests/test_post_processor.py#L304

**Comment by ruimashita** [comment]
@yd8534976 What license of the images and annotation jsons you added for test ? Can we use these images and json ? If you not sure about license, Please create json and image for test.

File: lmnet/tests/lmnet_tests/test_pre_processor.py (hunk 3 above)
Comment (ruimashita) [reply lmnet/tests/lmnet_tests/test_pre_processor.py]: @yd8534976 Thanks

**Comment by yd8534976** [comment]
@ruimashita  I only selected images and annotation jsons whose license is https://creativecommons.org/licenses/by-sa/2.0/

> You are free to:
Share — copy and redistribute the material in any medium or format
Adapt — remix, transform, and build upon the material
for any purpose, even commercially.

I think there is no problem.

File: lmnet/lmnet/datasets/base.py [hunk 4]
```diff
@@ -159,6 +159,45 @@ def _change_gt_boxes_shape(self, gt_boxes_list):
         return np.array(results)
 
 
+class KeypointsDetectionBase(Base, metaclass=ABCMeta):
```
Comment (tkng) [inline lmnet/lmnet/datasets/base.py]: On Google Scholar, "Key Point Detection" hits 5,340 documents, while "Key Points Detection" hits 917 documents. How about changing the class name to `KeypointDetectionBase`?

Also, according to the member method, this class exists for "single person pose estimation", hence "SinglePersonPoseEstimationBase" or something looks better class name.

Comment (yd8534976) [reply lmnet/lmnet/datasets/base.py]: @tkng , thanks!

I have checked mscoco official site: http://cocodataset.org/#keypoints-2019
As you said, `KeypointDetectionBase` is better.

For `SinglePersonPoseEstimationBase` name, `crop_from_full_image()` can also be used for `top-down approach multi-person pose estimation` or `facial keypoint (landmark) detection`. (Although we won't provide them.) `KeypointDetectionBase ` should be better here I think.

Comment (tkng) [reply lmnet/lmnet/datasets/base.py]: I see, thanks.

**Comment by tkng** [comment]
- [x] readability approval
- [x] ownership approval

**Pull request merged**